File: Rock.Blocks/Crm/AttributeValues.cs
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Net;

using Rock.Attribute;
using Rock.Data;
using Rock.Model;
using Rock.ViewModel.NonEntities;
using Rock.Web.Cache;

namespace Rock.Blocks.Crm
{
    /// <summary>
    /// Allows for editing the value(s) of a set of attributes for person.
    /// </summary>
    /// <seealso cref="Rock.Blocks.RockObsidianBlockType" />

    [DisplayName( "Attribute Values" )]
    [Category( "Obsidian > CRM" )]
    [Description( "Allows for editing the value(s) of a set of attributes for person." )]
    [IconCssClass( "fa fa-user" )]

    #region Block Attributes

    [AttributeCategoryField(
        "Category",
        Key = AttributeKey.Category,
        AllowMultiple = true,
        Description = "The Attribute Categories to display attributes from",
        EntityTypeName = "Rock.Model.Person",
        IsRequired = true,
        Order = 0 )]

    [TextField(
        "Attribute Order",
        Key = AttributeKey.AttributeOrder,
        Description = "The order to use for displaying attributes.  Note: this value is set through the block's UI and does not need to be set here.",
        IsRequired = false,
        Order = 1 )]

    [BooleanField(
        "Use Abbreviated Name",
        Key = AttributeKey.UseAbbreviatedName,
        Description = "Display the abbreviated name for the attribute if it exists, otherwise the full name is shown.",
        IsRequired = true,
        DefaultBooleanValue = false,
        Order = 2 )]

    [TextField( "Block Title",
        Key = AttributeKey.BlockTitle,
        Description = "The text to display as the heading.",
        IsRequired = false,
        DefaultValue = "",
        Order = 3 )]

    [TextField(
        "Block Icon",
        Key = AttributeKey.BlockIcon,
        Description = "The css class name to use for the heading icon.",
        IsRequired = false,
        DefaultValue = "",
        Order = 4 )]

    [BooleanField(
        "Show Category Names as Separators",
        Key = AttributeKey.ShowCategoryNamesAsSeparators,
        Description = "If enabled, attributes will be grouped by category and will include the category name as a heading separator.",
        IsRequired = true,
        DefaultBooleanValue = false,
        Order = 5 )]

    #endregion Block Attributes

    public class AttributeValues : RockObsidianBlockType
    {
        #region Attribute Keys
        private static class AttributeKey
        {
            public const string Category = "Category";
            public const string AttributeOrder = "AttributeOrder";
            public const string UseAbbreviatedName = "UseAbbreviatedName";
            public const string BlockTitle = "BlockTitle";
            public const string BlockIcon = "BlockIcon";
            public const string ShowCategoryNamesAsSeparators = "ShowCategoryNamesasSeparators";
        }

        #endregion Attribute Keys

        #region IRockObsidianBlockType Implementation

        /// <summary>
        /// Gets the property values that will be sent to the browser.
        /// </summary>
        /// <returns>
        /// A collection of string/object pairs.
        /// </returns>
        public override object GetObsidianBlockInitialization()
        {
            return new
            {
                BlockIconCssClass = GetBlockIconCssClass(),
                BlockTitle = GetBlockTitle(),
                ShowCategoryNamesAsSeparators = GetAttributeValue( AttributeKey.ShowCategoryNamesAsSeparators ).AsBoolean(),
                UseAbbreviatedNames = GetAttributeValue( AttributeKey.UseAbbreviatedName ).AsBoolean(),
                CategoryGuids = GetCategoryGuids(),
                Attributes = GetClientAttributeValues( RequestContext.GetContextEntity<Person>() )
            };
        }

        #endregion

        #region Actions

        /// <summary>
        /// Gets the attribute values for editing.
        /// </summary>
        /// <returns>A collection of attribute values that can be edited.</returns>
        [BlockAction]
        public BlockActionResult GetAttributeValuesForEdit()
        {
            var categoryGuids = GetCategoryGuids();
            var person = RequestContext.GetContextEntity<Person>();

            var attributeValues = person.GetClientEditableAttributeValues( RequestContext.CurrentPerson )
                .Where( av => categoryGuids.Count == 0 || categoryGuids.Any( g => av.Categories.FirstOrDefault( c => c.Guid == g ) != null ) )
                .ToList();

            return ActionOk( attributeValues );
        }

        /// <summary>
        /// Saves the attribute values.
        /// </summary>
        /// <param name="personGuid">The person unique identifier.</param>
        /// <param name="attributeValues">The attribute values.</param>
        /// <returns></returns>
        [BlockAction]
        public BlockActionResult SaveAttributeValues( Dictionary<string, string> keyValueMap )
        {
            using ( var rockContext = new RockContext() )
            {
                var currentPerson = RequestContext.CurrentPerson;

                var personService = new PersonService( rockContext );
                var person = personService.Get( RequestContext.GetContextEntity<Person>()?.Id ?? 0 );

                if ( person == null || currentPerson == null )
                {
                    return ActionNotFound();
                }

                person.LoadAttributes();
                var validKeys = GetCategoryAttributeKeys( person );

                foreach ( var kvp in keyValueMap )
                {
                    if ( validKeys.Contains( kvp.Key ) )
                    {
                        person.SetClientAttributeValue( kvp.Key, kvp.Value, currentPerson );
                    }
                }

                person.SaveAttributeValues( rockContext );

                return ActionOk( GetClientAttributeValues( person ) );
            }
        }

        /// <summary>
        /// Gets the attributes keys that are valid given our category configuration.
        /// </summary>
        /// <returns></returns>
        private List<string> GetCategoryAttributeKeys( IHasAttributes entity )
        {
            var categories = GetCategoryGuids();

            return entity.Attributes.Values
                .Where( a => a.Categories.Any( c => categories.Contains( c.Guid ) ) )
                .Select( a => a.Key )
                .ToList();
        }

        #endregion Actions

        #region Data Access

        /// <summary>
        /// Gets the block icon CSS class.
        /// </summary>
        /// <returns></returns>
        private string GetBlockIconCssClass()
        {
            var icon = GetAttributeValue( AttributeKey.BlockIcon );

            if (!icon.IsNullOrWhiteSpace())
            {
                return icon;
            }

            var category = GetSoloCategory();

            if ( category != null )
            {
                return category.IconCssClass;
            }

            return "fa fa-user";
        }

        /// <summary>
        /// Gets the block title.
        /// </summary>
        /// <returns></returns>
        private string GetBlockTitle()
        {
            var blockTitle = GetAttributeValue( AttributeKey.BlockTitle );

            if ( !blockTitle.IsNullOrWhiteSpace() )
            {
                return blockTitle;
            }

            var category = GetSoloCategory();

            if ( category != null )
            {
                return category.Name;
            }

            return "Attribute Values";
        }

        /// <summary>
        /// Gets the solo category.
        /// </summary>
        /// <returns></returns>
        private CategoryCache GetSoloCategory()
        {
            var categoryGuids = GetCategoryGuids();
            return categoryGuids.Count == 1 ? CategoryCache.Get( categoryGuids[0] ) : null;
        }

        /// <summary>
        /// Gets the attribute category Guids.
        /// </summary>
        /// <returns></returns>
        private List<Guid> GetCategoryGuids()
        {
            return GetAttributeValueAsFieldType<List<Guid>>( AttributeKey.Category );
        }

        /// <summary>
        /// Gets the client attribute values.
        /// </summary>
        /// <returns></returns>
        private List<ClientAttributeValueViewModel> GetClientAttributeValues( IHasAttributes entity )
        {
            var categoryGuids = GetCategoryGuids();

            return entity.GetClientAttributeValues( RequestContext.CurrentPerson )
                .Where( av => categoryGuids.Count == 0 || categoryGuids.Any( g => av.Categories.FirstOrDefault( c => c.Guid == g ) != null ) )
                .ToList();
        }

        #endregion Data Access
    }
}

File: Rock.JavaScript.Obsidian.Blocks/src/Crm/attributeValues.ts
// <copyright>
// Copyright by the Spark Development Network
//
// Licensed under the Rock Community License (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.rockrms.com/license
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>
//

import { Guid } from "@Obsidian/Types";
import { computed, defineComponent, ref } from "vue";
import Block from "@Obsidian/Templates/block";
import Loading from "@Obsidian/Controls/loading";
import NotificationBox from "@Obsidian/Controls/notificationBox.obs";
import { useStore } from "@Obsidian/PageState";
import { useConfigurationValues, useInvokeBlockAction } from "@Obsidian/Utility/block";
import JavaScriptAnchor from "@Obsidian/Controls/javaScriptAnchor";
import RockForm from "@Obsidian/Controls/rockForm";
import TextBox from "@Obsidian/Controls/textBox";
import RockButton from "@Obsidian/Controls/rockButton";
import { PublicAttributeBag } from "@Obsidian/ViewModels/Utility/publicAttributeBag";
import AttributeValuesContainer from "@Obsidian/Controls/attributeValuesContainer";
import { List } from "@Obsidian/Utility/linq";

const store = useStore();

type ConfigurationValues = {
    blockIconCssClass: string;

    blockTitle: string;

    showCategoryNamesAsSeparators: boolean;

    useAbbreviatedNames: boolean;

    categoryGuids: Guid[];

    attributes: Record<string, PublicAttributeBag>;

    values: Record<string, string>;
};

function sortedAttributeValues(attributeValues: PublicAttributeBag[]): PublicAttributeBag[] {
    return new List(attributeValues)
        .orderBy(v => v.order)
        .thenBy(v => v.name)
        .toArray();
}

export default defineComponent({
    name: "Crm.AttributeValues",
    components: {
        NotificationBox,
        Block,
        Loading,
        JavaScriptAnchor,
        RockForm,
        TextBox,
        RockButton,
        AttributeValuesContainer
    },
    setup() {
        const configurationValues = useConfigurationValues<ConfigurationValues>();
        const invokeBlockAction = useInvokeBlockAction();
        const attributes = ref(configurationValues.attributes);
        const attributeValues = ref(configurationValues.values);
        const personKey = computed(() => store.personContext?.idKey || null);
        const isLoading = ref(false);
        const isEditMode = ref(false);
        const errorMessage = ref("");

        const goToViewMode = (): void => {
            isEditMode.value = false;
        };

        const goToEditMode = async (): Promise<void> => {
            const result = await invokeBlockAction<PublicAttributeBag[]>("GetAttributeValuesForEdit");
            if (result.isSuccess) {
                //attributeValues.value = sortedAttributeValues(result.data ?? []);
                isEditMode.value = true;
            }
        };

        const doSave = async (): Promise<void> => {
            isLoading.value = true;

            const keyValueMap: Record<string, string | null> = {};

            //for (const a of attributeValues.value) {
            //    keyValueMap[(a as PublicEditableAttributeValue).key] = a.value || "";
            //}

            //const result = await invokeBlockAction<PublicAttributeValue[]>("SaveAttributeValues", {
            //    personKey: personKey.value,
            //    keyValueMap
            //});

            //if (result.isSuccess) {
            //    attributeValues.value = sortedAttributeValues(result.data ?? []);
            //    goToViewMode();
            //}
            //else {
            //    errorMessage.value = "Failed to save values.";
            //}

            isLoading.value = false;
        };

        return {
            blockTitle: computed(() => configurationValues.blockTitle),
            blockIconCssClass: computed(() => configurationValues.blockIconCssClass),
            isLoading,
            isEditMode,
            errorMessage,
            goToViewMode,
            goToEditMode,
            doSave,
            useAbbreviatedNames: configurationValues.useAbbreviatedNames,
            attributes,
            attributeValues
        };
    },
    template: `
<Block :title="blockTitle">
    <template #headerActions>
        <JavaScriptAnchor title="Order Attributes" class="action btn-link edit">
            <i class="fa fa-bars"></i>
        </JavaScriptAnchor>
        <JavaScriptAnchor title="Edit Attributes" class="action btn-link edit" @click="goToEditMode">
            <i class="fa fa-pencil"></i>
        </JavaScriptAnchor>
    </template>

    <template #default>
        <Loading :isLoading="isLoading">
            <NotificationBox v-if="errorMessage" alertType="warning">{{ errorMessage }}</NotificationBox>
            <AttributeValuesContainer v-if="!isEditMode" :attributeValues="attributeValues" :showEmptyValues="false" :showCategoryLabel="false" />
            <RockForm v-else @submit="doSave">
                <AttributeValuesContainer v-model="attributeValues" :attributes="attributes" isEditMode :showAbbreviatedName="useAbbreviatedNames" :showCategoryLabel="false" />
                <div class="actions">
                    <RockButton btnType="primary" btnSize="xs" type="submit">Save</RockButton>
                    <RockButton btnType="link" btnSize="xs" @click="goToViewMode">Cancel</RockButton>
                </div>
            </RockForm>
        </Loading>
    </template>
</Block>`
});
